sympy/core/basic.py

```diff
@@ -1077,6 +1077,151 @@ def _eval_subs(self, old, new):
         """
         return None
 
+    def dsubs(self, subsdict, mapdict=None):
+        r"""
+        Function that substitutes for variables in an ordinary differential equation.
+
+        The inputs are ``subsdict`` and ``mapdict``.
+
+        ``subsdict``: A dictionary specifying the variable transformations with
+            the old variable as key and the transformation as value.
+
+        ``mapdict``: In case a transformation contains more than one new
+        variable, mapdict must be provided to tell SymPy which variable
+        should be considered for substitution. For eg:
+
+        If you want to substitute for `x = at^2` where `t` is the new
+        independent variable, `subsdict` must be {x: a*t**2} and mapdict
+        must be {x: t}. It is ``None`` by default.
```

Comment:
I think maybe there can be a better way to define the signature so that mapdict is not needed.

Reply:
I couldn't find a way to do it. How do you think it should be done?

Reply:
How does maple resolve this?

Reply:
Maple has a parameter `params` which is a list of symbols present in the transformation rules that have to be treated as constants and not variables. Should I modify the API and include this instead of `mapdict`? Just as an example -
``` python
>>> ode = diff(f(x), x, x) + cos(x)
>>> ode.dsubs({x: a*t}, params=[a])
f(x).diff(x, 2)/a**2 + f(t)
```

Reply:
I think we should require the user to be explicit about what are the dependent and independent variables for the transformation. Maybe something like this:
```python
dsubs(eq, {f(x)}, {g(t)}, {x: a*t, f(x): g(t)})
```
Here we know that `t` is the new independent variable because it is listed as such in the second argument. Any function not listed could just have the independent variable change if appropriate so e.g. `h(x)` would just become `h(a*t)`.

A transformation of only the dependent variable could be like:
```python
dsubs(eq, {x(t)}, {y(t)}, {x(t): exp(-t)*y(t)})
```

Then if we have a PDE it can be:
```python
dsubs(eq, {f(x,y)}, {g(xi,eta)}, {x: xi + eta, y: xi - eta, f(x,y): g(xi,eta)})
```
Maple's idea of a unknown function could correspond fo giving `f(x):f(t)` (use the same function `f` but change the independent variable).

A system of ODEs could transform like e.g.:
```python
transformation = {t: exp(tau), x(t): a(tau) + b(tau), y(t): a(tau) - b(tau)}
dsubs(eqs, {x(t), y(t)}, {a(tau), b(tau)}, transformation)
```
So the syntax is:
```python
dsubs(expressions, old_dependent_vars, new_dependent_vars, transformation_equations)
```

I'm just trying to think if it would ever make sense to use `dsubs` in a situation where there are no "dependent variables"...

Reply:
I suppose it would still be useful if there was a function that sympy just didn't know how to differentiate like:
```python
In [7]: Derivative(zeta(2*t), t).subs(t, z/2)
Out[7]: 
⎛d         ⎞│   
⎜──(ζ(2⋅t))⎟│  z
⎝dt        ⎠│t=─
               2
```

Reply:
So maybe the signature could be:
```python
dsubs(expressions, old, new, transformation)
```
where old and new give both the dependent and independent variables.

For example:
```python
dsubs(eq, {f(x), x}, {g(t), t}, {x: exp(t), f(x): g(t)})
```

Reply:
I suppose that if we were going to make any of the arguments optional it would be `old` and `new` so we could put those last like:
```python
dsubs(expressions, transformation, old=None, new=None)
```
Then these could be equivalent:
```python
dsubs(eq, {x: exp(t), f(x): g(t)})
dsubs(eq, {x: exp(t), f(x): g(t)}, {f(x), x}, {g(t), t})
```
Since there are no other symbols or functions in the transformation equations it would be unambiguous in some cases.

Actually perhaps `old` isn't needed if everything in `old` is always required to be on the lhs in the dict. But then I'm not sure if we would want to allow the transformation equations to be given in a non-explicit form later...

Reply:
If we have any of `old`/`new` the other set can be easily obtained, so I don't think both are required. If both of them are given, we would have to again verify that they indeed match. I feel like the Maple syntax has lesser user input and is easier. The syntax you proposed would make the user specify a lot of things which ideally can be computed.

Reply:
I've changed the function signature to make it Maple like. Please review @oscarbenjamin

Reply:
> The syntax you proposed would make the user specify a lot of things which ideally can be computed.

Yes, but they are things that the caller of this function would already have so computing them is inefficient.

> I feel like the Maple syntax has lesser user input and is easier.

The maple syntax is nice for simple cases but not very good for more complex cases. It's also not easy to explain or understand. The difference between "known" and "unknown" functions doesn't make much sense to me (perhaps there is a better name to describe it).

Reply:
> The difference between "known" and "unknown" functions doesn't make much sense to me.

These examples will be helpful to understand `known` and `unknown` functions -
https://github.com/sympy/sympy/blob/f74e2eb22db6a69002b1cda2040e82166ec48032/sympy/core/basic.py#L1128-L1138

> Perhaps there is a better name to describe it.

Yes, we could rename them, but I couldn't think of a suitable name.